util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,190 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Executes a number of steps in a startup/shutdown process.
+ *
+ * <p>On startup, steps are executed in the given order. If any step completes exceptionally, then
+ * the subsequent steps are not executed and the startup future completes exceptionally. However, no
+ * shutdown is triggered by this class. This can be done by the caller.
+ *
+ * <p>On shutdown, steps are executed in reverse order. If any shutdown step completes
+ * exceptionally, subsequent steps will be executed and the exceptions of all steps are collected as
+ * suppressed exceptions
+ *
+ * <p>Callers of this class must obey the following contract:
+ *
+ * <ul>
+ *   <li>Shutdown must not be called before startup
+ *   <li>Startup must be called at most once
+ *   <li>Shutdown may be called more than once. The first call will trigger the shutdown and any
+ *       subsequent calls do nothing
+ *   <li>Shutdown may be called before the future of startup has completed. In that case, it will
+ *       complete the current running startup step, cancel all subsequent startup step, complete the
+ *       startup future with an exception and start the shutdown from the step that last completed
+ * </ul>
+ *
+ * @param <CONTEXT> the startup/shutdown context
+ */
+public final class StartupProcess<CONTEXT> {
+
+  private final List<StartupStep<CONTEXT>> steps;
+  private final String processName;
+  private final Logger logger;
+
+  private boolean startupCalled = false;
+  private final Stack<StartupStep<CONTEXT>> startedSteps = new Stack<>();
+  private CompletableFuture<CONTEXT> shutdownFuture;
+
+  public StartupProcess(final String processName, final List<StartupStep<CONTEXT>> steps) {
+    this.steps = Collections.unmodifiableList(Objects.requireNonNull(steps));
+
+    this.processName = processName != null ? processName : "Undefined process";
+    logger =
+        LoggerFactory.getLogger(
+            "io.camunda.zeebe.util.startup.StartupProcess(" + this.processName + ")");
+  }
+
+  /**
+   * Executes the startup logic
+   *
+   * @param context the startup context at the start of this step
+   * @return future with startup context at the end of this step
+   */
+  public synchronized CompletableFuture<CONTEXT> startup(final CONTEXT context) {
+    logger.debug("Startup was called with context: " + context);
+    if (startupCalled) {
+      throw new IllegalStateException("startup(...) must only be called once");
+    }
+    startupCalled = true;
+
+    final var startupFuture = new CompletableFuture<CONTEXT>();
+    final var stepsToStart = new ArrayList<>(steps);
+
+    logger.info("Starting startup process");
+    proceedWithStartup(stepsToStart, context, startupFuture);
+
+    return startupFuture;
+  }
+
+  private void proceedWithStartup(
+      final ArrayList<StartupStep<CONTEXT>> stepsToStart,
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> startupFuture) {
+    if (shutdownFuture != null) {
```

Comment:
:x: I think there's a potential race condition here if `shutdownFuture` is not volatile. I could be wrong though, so bear with me :smile: `startup` and `shutdown` are both synchronized, so they cannot interfere with each other - however, `proceedWithStartup` is not, so it may be called, and then `shutdown` is called, and then the first step is completed, which triggers a new `proceedWithStartup` call which is not necessarily called from the same thread as `shutdown`. In this case, it's possible that here we don't see the latest value of `shutdownFuture`, resulting in us continuing with the startup process in parallel with the shutdown process. wdyt?

This is where (and I know you'll hate me for it :smile:) having these things be actors would simplify things (for this case - it would arguably complicate other things :smile:)